HW0/P4/P4.py

```diff
@@ -40,37 +44,92 @@
     # Read the observation array and plot it (Part 2)
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    measured_inputs = np.loadtxt('P4_measurements.txt', delimiter=',')
+    stretch = np.array([[1.0/rx, 0.0, 0.0],[0.0, 1.0/ry, 0.0],[0.0, 0.0, 1.0/rz]])
+    stretched_inputs = np.dot(stretch, measured_inputs.T)
+    x_coords = stretched_inputs[0,:]
+    y_coords = stretched_inputs[1,:]
+    z_coords = stretched_inputs[2,:]
+    ax.plot(x_coords, y_coords, z_coords,
+            '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
-
     # Initial conditions for s0
-    # Compute the rest of sk using Eq (1)
+    A = np.array([[1.0, 0.0, 0.0, dt, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0, dt, 0.0], [0.0, 0.0, 1.0, 0.0, 0.0, dt],
+                  [0.0, 0.0, 0.0, (1.0 - (c*dt)), 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, (1.0 - (c*dt)), 0.0], 
+                  [0.0, 0.0, 0.0, 0.0, 0.0, (1.0 - (c*dt))]])
+    a = np.array([0.0, 0.0, 0.0, 0.0, 0.0, g*dt]).T
+    a.shape = (6,1)
+    s_old = np.array([0.0, 0.0, 2.0, 15.0, 3.5, 4.0]).T
+    s_old.shape = (6,1)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Compute the rest of sk using Eq (1)
+    outputs = s_old
+    for k in range(0, K):
+        s_new = np.dot(A, s_old) + a
+        s_new.shape = (6,1)
+        outputs = np.append(outputs, s_new, axis=1)
+        s_old = s_new
+    
+    x_coords = outputs[0,:]
+    y_coords = outputs[1,:]
+    z_coords = outputs[2,:]
+
+    ax.plot(x_coords, y_coords, z_coords,
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.array([[bx, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, by, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, bz, 0.0, 0.0, 0.0],
+                  [0.0, 0.0, 0.0, bvx, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, bvy, 0.0],
+                  [0.0, 0.0, 0.0, 0.0, 0.0, bvz]])
+    C = np.array([[rx, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, ry, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, rz, 0.0, 0.0, 0.0]])
 
     # Initial conditions for s0 and Sigma0
+    s_old = np.array([0.0, 0.0, 2.0, 15.0, 3.5, 4.0]).T
+    s_old.shape = (6,1)
+    
+    sig_old = 0.01 * np.identity(6)
+    
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    def predictS (sk):
+        guess = np.dot(A, sk) + a
+        guess.shape = (6,1)
+        return guess
+
+    def predictSig (sigk):
+        return np.linalg.inv(np.dot(np.dot(A, sigk) , A.T) + np.dot(B, B.T))
```

Comment:
Same thing here, instead of using np.dot(A,B) you can use A.dot(B) for simplicity.